### ISSUE TITLE
Can't use my MariaDB DB

### ISSUE THREAD

**Issue by romu70** [wallabag/docker#9]
Hi,
First of all, I'm far from being a Docker expert, sorry if this is a newbie issue.

I'm trying to re-build the Docker image to fits my own needs, in particular, to run Wallabag on my Synology, through a Docker container.
1. I checked the Dockerfile works "out of the box", it does indeed
2. I checked on my laptop, I can connect my Synology MariaDB with the "wallabag" credentials I created, it works

Then, I changed the Dockerfile like this:

```
ENV WALLABAG_VERSION=2.0.4 \
    SYMFONY__ENV__DATABASE_DRIVER=pdo_mysql \
    SYMFONY__ENV__DATABASE_HOST=<my synology IP address> \
    SYMFONY__ENV__DATABASE_PORT=3306 \
    SYMFONY__ENV__DATABASE_NAME=wallabag \
    SYMFONY__ENV__DATABASE_USER=wallabag \
    SYMFONY__ENV__DATABASE_PASSWORD=<the password I assigned to the wallabag user> \
    SYMFONY__ENV__SECRET=ovmpmAWXRCabNlMgzlzFXDYmCFfzGv
```

This is the only change I made to the Dockerfile, all the rest remained unchanged.

This way, `docker run` still works, but obviously, it still uses sqlite, because the MariaDB Wallabag DB on the Syno stays totally empty.

Any idea of what I did wrong? Thanks.

**Comment by McMac** [comment]
Trying the exact same thing and from what I can read the ansible/entrypoint.yml does not support connecting to an existing/pre-created dbs using a given user. Always wants to use root to set it all up. Unfortunately can't fix it or propose a fix.
I do get a access denied for 

> TASK [add mariadb db] **********************************************************
> fatal: [localhost]: FAILED! => Access denied for user 'root'@rancher..." 

i.e. it is not picking up the user given nor the host provided.

**Comment by xsteadfastx** [comment]
@romu70 the problem is that the Dockerfile is only used when you create the image. The run command just starts a container from a already created image. To use MariaDB or something else you have to give the docker run command environment variables and when you spin up the container it will check for the environment variables and use the MariaDB instead of sqlite. For example:

`$ docker run --name wallabag --link wallabag-db:wallabag-db -e "MYSQL_ROOT_PASSWORD=my-secret-pw" -e "SYMFONY__ENV__DATABASE_DRIVER=pdo_mysql" -e "SYMFONY__ENV__DATABASE_HOST=wallabag-db" -e "SYMFONY__ENV__DATABASE_PORT=3306" -e "SYMFONY__ENV__DATABASE_NAME=wallabag" -e "SYMFONY__ENV__DATABASE_USER=wallabag" -e "SYMFONY__ENV__DATABASE_PASSWORD=wallapass" -p 80:80 wallabag/wallabag`

This will overwrite the defaults set in the dockerfile aka the image. If you just do a `docker run ...` it will use the defaults... sqlite.

**Comment by romu70** [comment]
Ok, thanks @xsteadfastx . So, to make my changes "persistent", I've to first run ```docker build...```, instead of ```run```, right?

**Comment by xsteadfastx** [comment]
i think the best way is to create a environment file. for example `env.mariadb` with this content:

```
MYSQL_ROOT_PASSWORD=wallaroot
SYMFONY__ENV__DATABASE_DRIVER=pdo_mysql
SYMFONY__ENV__DATABASE_HOST=db
SYMFONY__ENV__DATABASE_PORT=3306
SYMFONY__ENV__DATABASE_NAME=wallabag
SYMFONY__ENV__DATABASE_USER=wallabag
SYMFONY__ENV__DATABASE_PASSWORD=wallapass
```

and now you can start the docker container with `docker run --env-file=env.mariadb -p 80:80 wallabag/wallabag` and your settings are always the same.

**Comment by romu70** [comment]
I'll try this on the coming days, thanks.

**Comment by McMac** [comment]
@romu70: docker run = docker build (image) + docker start (container)

@xsteadfastx: 
- good idea on the env file will need to be transferred to the docker host but should make the call easier (cut'n paste of the command line vs. transferring the file (vi + cut'n paste))
- if the database and user has been created via a phpMyAdmin e.g. is MYSQL_ROOT_PASSWORD still necessary?
- the call needs to be modified to not include --link as when we are using an external non-docker DB we don't link wallabag with any db docker container

**Comment by xsteadfastx** [comment]
@McMac i dont have much time right now but if i have some minutes i try to get my head around all this :) usually the entrypoint just needs the root user and password if the db is not there yet. else it will skip the creation... and doesnt need any root user or something. we also could add another environment variable for a db admin user... if its not root.

**Comment by McMac** [comment]
@xsteadfastx thanks! I have never used ansible so I can't propose a fix at this point. My simple (?) idea would be: if no MYSQL_ROOT_PASSWORD has been provided simply assume SYMFONY__ENV... points to an existing DB and user and therefore skip creation of either. I believe if one is accessing wallabag the first time the tables themselves get created?  
I had provided only (actually with and without MYSQL_ROOT_PWD):
'''
SYMFONY__ENV__DATABASE_DRIVER=pdo_mysql
SYMFONY__ENV__DATABASE_HOST=db
SYMFONY__ENV__DATABASE_PORT=3306
SYMFONY__ENV__DATABASE_NAME=wallabag
SYMFONY__ENV__DATABASE_USER=wallabag
SYMFONY__ENV__DATABASE_PASSWORD=wallapass
'''
and then received the error:

> TASK [add mariadb db] **********************************************************
> fatal: [localhost]: FAILED! => Access denied for user 'root'@rancher..." 

-> different user and different DB host than specified...

**Comment by McMac** [comment]
Ok, got it working. Create entry for root to allow to remote connect, and remove manually created wallabag user + db (looks like if the latter two exist then wallabag won't create table structure). Then re-run docker run...

**Comment by xsteadfastx** [comment]
@McMac so its about using a existing mysql server and not a fresh mysql standalone container?

**Comment by McMac** [comment]
Correct. Existing mysql server - not necessarily - new database. 

Use case is:

1. As I use a low-power ATOM based server, load-balance between Web server (docker) and DB server somewhere else (my case, Synology box)
2. Have 1 DB server for internal and external use -> backup 
3. To no have to throw around root passwords (DB or the like), create database and user via script on the DB server directly, wallabag to detect: 
- wallabag credentials provided and connect works -> use DB (no need for root to create user and DB)
- DB no tables exist (e.g. wallabag user table) then create tables using the wallabag user (or provide schema to be created along the database and user)

Depending on how wallabag internally creates the DB structure, either expose the same script or do the detecting thing and create then (via flag to only have to do this once - URL param?). 

Bottom line: no root pwd - no create

**Comment by xsteadfastx** [comment]
sorry im so slow. i have alot of work to do. but i will see what i can do about it. maybe i just need some more variables for setting an admin user to use something like that. and honestly i never thought about using a already exisiting, not containerized database for storing the data. thats why i didnt test it.

**Comment by McMac** [comment]
Well, actually it's just about connecting to an existing DB or create from scratch compared to whether it is dockerized or not. 
If you only want to run wallabag from scratch and only that, then this is fine. Start adding other webapps also running on MySQL then chances are you won't fire up new containers for it...

**Comment by McMac** [comment]
Attempting to give it another try however can't figure it out how to do it. If it is supported then please describe the steps.

**Comment by Thrilleratplay** [comment]
I haven't fully tested this as a script but these are the commands I used to set up my Wallabag, Redis and MariaDB docker instances

https://gist.github.com/Thrilleratplay/ed0c74024cfc6ff52d2f727f49ca6925

**Comment by McMac** [comment]
Thanks for sharing. I have it running however the Redis doesn't seem to be processing anything.

**Comment by xsteadfastx** [comment]
did it worked? can we close this?

**Comment by romu70** [comment]
yes we can (oups, obsolete now!)!
